qemu/tests/qemu_guest_agent.py

```diff
@@ -507,9 +516,42 @@ def gagent_check_powerdown(self, test, params, env):
         :param params: Dictionary with the test parameters
         :param env: Dictionary with test environmen.
         """
-        self.__gagent_check_shutdown(self.gagent.SHUTDOWN_MODE_POWERDOWN)
-        if not utils_misc.wait_for(self.vm.is_dead, self.vm.REBOOT_TIMEOUT):
-            test.fail("Could not shutdown VM via guest agent'")
+        def _gagent_check_shutdown(self):
+            self.__gagent_check_shutdown(self.gagent.SHUTDOWN_MODE_POWERDOWN)
+            if not utils_misc.wait_for(self.vm.is_dead, self.vm.REBOOT_TIMEOUT):
+                test.fail("Could not shutdown VM via guest agent'")
+
+        session = self._get_session(params, self.vm)
+        self._open_session_list.append(session)
+
+        if params.get("os_type") == "linux":
+            self.log_persistence(params, session, self.vm)
+            _gagent_check_shutdown(self)
+
+            time.sleep(20)
+            env_process.preprocess_vm(test, params, env, params["main_vm"])
+            vm = env.get_vm(params["main_vm"])
+            login_timeout = int(params.get("login_timeout", 360))
+            session = vm.wait_for_login(timeout=login_timeout)
+
+            start_vm = self.params["start_vm"]
+            self.start_vm = start_vm
+            if self.start_vm == "yes":
+                vm.verify_alive()
+                self.vm = vm
```

Comment:
```suggestion
            self.vm = env.get_vm(params["main_vm"])
            session = self.vm.wait_for_login(timeout=int(params.get("login_timeout", 360)))
```

Reply:
Step vm.verify_alive() should be passed and then self.vm = vm , isn't it? sorry ,I don't know I understand your suggestion correctly or not, you mean just reserve following two lines and delete original 533~541 lines right?:

self.vm = env.get_vm(params["main_vm"])
session = self.vm.wait_for_login(timeout=int(params.get("login_timeout", 360)))

Reply:
> Step vm.verify_alive() should be passed and then self.vm = vm , isn't it? sorry ,I don't know I understand your suggestion correctly or not, you mean just reserve following two lines and delete original 533~541 lines right?:
> 
> self.vm = env.get_vm(params["main_vm"])
> session = self.vm.wait_for_login(timeout=int(params.get("login_timeout", 360)))
1. You have `wait_for_login` in L535, which include `verify_alive`, so it's not necessary to verify it again.
2. Yes that's what I mean, only this two line is enough, if it is not alive you will fail on `wait_for_login`.